cmdstanpy/utils.py

```diff
@@ -1061,3 +1061,74 @@ def __enter__(self):
     def __exit__(self, exc_type, exc_val, exc_tb):
         if self._tmpdir:
             shutil.rmtree(self._tmpdir, ignore_errors=True)
+
+
+def extract(
+    fit: Any,
+    parameters: Optional[List] = None,
+    inc_diagnostics: Optional[bool] = False,
+    inc_warmup: Optional[bool] = False,
+) -> Dict[str, Any]:
+    """Extract ndim dictionary from CmdStanPy fit.
+
+    Parameters
+    ----------
+    fit: CmdStanMCMC
+    parameters: list
+    inc_diagnostics: bool
+        Don't include diagnostics with default paramaters.
+    inc_warmup: bool
+
+    Returns
+    -------
+    dict
+    """
+    if parameters is None:
+        parameters = fit.column_names
+        if not inc_diagnostics:
+            parameters = [
+                param for param in parameters if not param.endswith("__")
+            ]
+
+    if inc_warmup and not fit._save_warmup:
+        inc_warmup = False
+
+    data = fit.draws(inc_warmup=inc_warmup)
+    draws, chains, *_ = data.shape
+    column_groups = defaultdict(list)
+    column_locs = defaultdict(list)
+    # iterate flat column names
+    for i, col in enumerate(fit.column_names):
+        col_base, *col_tail = col.replace(']', '').replace('[', ',').split(',')
+        if col_tail:
+            # gather nD array locations
+            column_groups[col_base].append(tuple(map(int, col_tail)))
+        # gather raw data locations for each parameter
+        column_locs[col_base].append(i)
+    dims = {}
+    for colname, col_dims in column_groups.items():
+        # gather parameter dimensions (assumes dense arrays)
+        dims[colname] = tuple(np.array(col_dims).max(0))
```

Comment:
this is equivalent to utility function `parse_var_dims`?   does this handle multi-dim arrays?
e.g., Stan variable that is an array of matrices?
```
int d;
  matrix [d,d] d_m;
  matrix [d,d] d_m_ar[d];
```

which version is more efficient?

as per comment - this only works on dense arrays